ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java

```diff
@@ -488,40 +494,76 @@ public void close() {
+      Map<StoreKey, IndexFinalState> keyFinalStates, boolean removeFromStates, IndexEntryAction validIndexEntryAction)
+      throws StoreException {
     ListIterator<IndexEntry> it = indexSegment.listIterator(indexSegment.size());
     while (it.hasPrevious()) {
       IndexEntry indexEntry = it.previous();
       IndexValue indexValue = indexEntry.getValue();
+      StoreKey key = indexEntry.getKey();
       if (indexValue.isDelete()) {
-        // delete record is always valid
+        if (keyFinalStates.containsKey(key)) {
+          IndexFinalState state = keyFinalStates.get(key);
+          if (state.isUndelete() || (state.isDelete() && state.getLifeVersion() != indexValue.getLifeVersion()) || state
+              .isTtlUpdate()) {
+            // This DELETE is not valid, when the final state of this storeKey is
+            // 1. UNDELETE, or
+            // 2. DELETE, but the current lifeVersion is not the same, or
+            // 3. TTL_UPDATE
+            continue;
+          }
+        } else {
+          long operationTimeInMs =
+              indexValue.getOperationTimeInMs() == Utils.Infinite_Time ? indexSegment.getLastModifiedTimeMs()
+                  : indexValue.getOperationTimeInMs();
+          keyFinalStates.put(indexEntry.getKey(),
+              new IndexFinalState(indexValue.getFlags(), operationTimeInMs, indexValue.getLifeVersion()));
+        }
         validIndexEntryAction.accept(indexEntry);
-        if (!isExpired(indexValue.getExpiresAtMs(), referenceTimeInMs)) {
+      } else if (indexValue.isUndelete()) {
+        if (keyFinalStates.containsKey(key)) {
+          IndexFinalState state = keyFinalStates.get(key);
+          if (state.isTtlUpdate()) {
+            indexValue.setExpiresAtMs(Utils.Infinite_Time);
+          }
+          if (state.isDelete() || (state.getLifeVersion() != indexValue.getLifeVersion())) {
+            // This UNDELETE is not valid, when the final state of this storeKey is
+            // 1. DELETE, or
+            // 2. the current lifeVersion is not the same
+            continue;
+          }
+        } else {
           long operationTimeInMs =
               indexValue.getOperationTimeInMs() == Utils.Infinite_Time ? indexSegment.getLastModifiedTimeMs()
                   : indexValue.getOperationTimeInMs();
-          deletedKeys.put(indexEntry.getKey(), operationTimeInMs);
+          keyFinalStates.put(indexEntry.getKey(),
+              new IndexFinalState(indexValue.getFlags(), operationTimeInMs, indexValue.getLifeVersion()));
+        }
+        if (!isExpired(indexValue.getExpiresAtMs(), referenceTimeInMs)) {
+          validIndexEntryAction.accept(indexEntry);
         }
       } else if (indexValue.isTtlUpdate()) {
-        if (isTtlUpdateEntryValid(indexEntry.getKey(), indexValue, referenceTimeInMs, deletedKeys)) {
+        if (isTtlUpdateEntryValid(key, indexValue, referenceTimeInMs, keyFinalStates)) {
           validIndexEntryAction.accept(indexEntry);
         }
```

Comment:
Maybe I overlooked something but I feel like it doesn't check if final state is delete case when validating the current TtlUpdate, does it? Please point me to the logic if I missed anything.

Reply:
it's in line 595. It checks if there is a put in the same log segment with ttl update, if so, check if this put is valid or not. if it's not, then the ttl update is not valid either.

Reply:
I see, what if put is not in the same log segment, how do we handle that?

Reply:
There is a comments regarding this case. As long as the PUT exists and not in the same log segment, TTL_UPDATE is considered as valid. I suppose this is because the stats is used to trigger compaction, and it's possible that in compaction we only compact log segment with TTL_UPDATE and not PUT. For instance, if PUT is in 0_0_log and TTL_UPDATE is in 1_0_log and in compaction, we only compaction 1_0_log, in this case, TTL_UPDATE would be considered as valid and will be kept after compaction. I suppose the logic here is trying to be as close to the compaction logic as possible.